tic-tac-toe/src/app/screens/Game/index.js

```diff
@@ -13,44 +17,36 @@ function calculateWinner(squares) {
   return winner;
 }
 class GameContainer extends Component {
-  state = {
-    history: [
-      {
-        squares: Array(9).fill(null)
-      }
-    ],
-    stepNumber: 0,
-    xIsNext: true
-  };
-
   handleClick = i => {
-    const history = this.state.history.slice(0, this.state.stepNumber + 1);
+    const history = this.props.history.slice(0, this.props.stepNumber + 1);
     const current = history[history.length - 1];
     const squares = current.squares.slice();
     if (calculateWinner(squares) || squares[i]) {
       return;
     }
-    squares[i] = this.state.xIsNext ? 'X' : 'O';
-    this.setState({
+    squares[i] = this.props.xIsNext ? 'X' : 'O';
+    const state = {
       history: history.concat([
         {
           squares
         }
       ]),
       stepNumber: history.length,
-      xIsNext: !this.state.xIsNext
-    });
+      xIsNext: !this.props.xIsNext
```

Comment:
You should't change the state here. The reducer should change stuff like who goes next and the history of the moves.